File: binance-dart/lib/websocket_handler.dart
import 'dart:io';
import 'dart:convert';

class WebsocketHandler {
  final String _taskName;
  final void Function(Map<String, dynamic>) _handleData;
  Set<String> _subscriptions = Set<String>();

  final int _reconnectTime = 5;
  WebSocket? _ws;
  bool _keepConnected = true;
  late String _baseUrl;

  WebsocketHandler(this._taskName, this._handleData, {useTestnet = false}) {
    _baseUrl = useTestnet
        ? 'wss://testnet.binance.vision/ws'
        : 'wss://stream.binance.com:9443/ws';
  }

  /// Starts the WebsocketHandler object.
  Future<void> start() async {
    while (_keepConnected) {
      try {
        // connects to websocket server
        print('$_taskName: connecting to $_baseUrl');
        _ws = await WebSocket.connect(_baseUrl);

        // subcribes to the subscriptions topics
        sendSubscription(_subscriptions.toList(), true);

        // receives messages
        print('$_taskName: waiting for messages');
        await for (var message in _ws!) {
          // converts the message to json
          var json = jsonDecode(message);

          // pass the message to the callback handler
          _handleData(json);
        }

        // if here, websocket was disconnected
        print('$_taskName: disconnected');
        _ws = null;
      } on SocketException catch (e) {
        // log an error message
        print(
            '$_taskName: Error ${e.osError?.errorCode} | ${e.osError?.message.trimRight()}');

        // removes reference
        _ws = null;

        // waits some time before retrying
        print('$_taskName: reconnecting in $_reconnectTime s');
        await Future.delayed(Duration(seconds: _reconnectTime));
      }
    }
  }

  // Disposes the WebsocketHandler resources.
  void dispose() {
    print('$_taskName: disposing resources');
    _keepConnected = false;
    _ws?.close();
  }

  /// Sends a subscribe / unsubscribe message over websocket.
  sendSubscription(List<String> streams, bool isSubscribe) {
    if (_ws != null && _ws!.readyState == WebSocket.open) {
      var payload = jsonEncode({
        "method": isSubscribe ? 'SUBSCRIBE' : 'UNSUBSCRIBE',
        "params": streams,
        "id": DateTime.now().toUtc().millisecondsSinceEpoch,
      });
      _ws!.add(payload);
    } else {
      print('$_taskName: ERROR: unable to send message over websocket');
    }
  }

  /// Subscribes to User Data stream.
  subscribeToUserData(String listenKey) {
    // add item to the subscription list
    _subscriptions.add(listenKey);

    // send a live subscription in the websocket if connected
    if (_ws != null && _ws!.readyState == WebSocket.open) {
      sendSubscription([listenKey], true);
    }
  }

  // /// forms the url to connect to a Mini Ticker Stream.
  // void setMiniTickerUrl(List<String> symbols) {
  //   var streams = symbols.map((s) => s.toLowerCase() + '@miniTicker');
  //   var urlBuffer = StringBuffer(_baseUrl)
  //     ..write('/stream?streams=')
  //     ..writeAll(streams, '/');
  //   _url = urlBuffer.toString();
  // }

  // /// forms the url to connect to a all Mini Tickers Stream.
  // void setMiniTickerAllUrl() {
  //   _url = '$_baseUrl/ws/!miniTicker@arr';
  // }

  // /// forms the url to connect to account stream.
  // void setAccountUrl(String listenKey) {
  //   _url = '$_baseUrl/ws';
  //   // _url = '$_baseUrl/stream';
  // }
}

File: binance-ts/src/websocket.handler.ts
import { connected } from 'process';
import { WebSocket } from 'ws';

type MessageHandler = (msg: Map<string, any>) => void;

export class WebsocketHandler {
	private _closedByError = false;
	private _baseUrl: string;
	private readonly _reconnectTime = 5000; // miliseconds
	private _ws?: WebSocket = undefined;
	private _subscriptions = new Set<string>();

	constructor(private readonly taskName: string, private callback: MessageHandler, testnet = false) {
		this._baseUrl = testnet ? 'wss://testnet.binance.vision/ws' : 'wss://stream.binance.com:9443/ws';

		// starts the socket
		this.start();
	}

	start() {
		// creates the websocket and connects it
		console.log(`${this.taskName}: connecting to ${this._baseUrl}`);
		this._ws = new WebSocket(this._baseUrl);

		// configures websocket connection
		this._ws.onopen = () => {
			console.log(`${this.taskName}: websocket connected`);
			this._closedByError = false;
			this._sendSubscription(Array.from(this._subscriptions), true);
		};

		// configures websocket closure
		this._ws.onclose = () => {
			console.log(`${this.taskName}: websocket closed`);
			this._ws = undefined;
			if (this._closedByError) {
				console.log(`${this.taskName}: retrying in ${this._reconnectTime / 1000}s ...`);
				setTimeout(() => this.start(), this._reconnectTime);
			} else {
				this.start();
			}
		};

		// configures websocket error
		this._ws.onerror = (err) => {
			console.log(`${this.taskName}: websocket error: ${err.message}`);
			this._closedByError = true;
			this._ws = undefined;
		};

		// configures websocket message handling
		this._ws.onmessage = (msg) => {
			// parse the received message as json
			var json: Map<string, any> = JSON.parse(msg.data as string);

			// pass the massage to the callback handler
			this.callback(json);
		};
	}

	private _sendSubscription(streams: string[], isSubscribe: boolean) {
		if (this._ws) {
			const payload = JSON.stringify({
				method: isSubscribe ? 'SUBSCRIBE' : 'UNSUBSCRIBE',
				params: streams,
				id: Date.now(),
			});
			console.log(payload);
			this._ws.send(payload);
		} else {
			console.log('ERROR: websocket not able to subscribe');
		}
	}

	subscribeToUserData(listenKey: string) {
		// add item to the subscription list
		this._subscriptions.add(listenKey);

		// send a live subscription in the websocket if connected
		if (this._ws && this._ws.readyState == WebSocket.OPEN) {
			this._sendSubscription([listenKey], true);
		}
	}
}
